### PR TITLE
Ana Lisa & Jamila: Octos-C9

### PR DESCRIPTION
**Pull request by The-Beez-Kneez**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the advantages of using git when collaboratively working on one code base? | It is easy for people to consider the problem separately on different screens. 
Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? | So that the overall module Scrabble would be able to access Scoring Class elsewhere in the program
Describe an example of a test you wrote for a _nominal case_. | We tested whether or not the plays method could return an array.
Describe an example of a test you wrote for an _edge case_. | We had some drafted up for won but did not get the chance to implement them.
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | We used inject to receive the total_score of our player's word scores.
What was one method you and your pair used to debug code? | We utilized pry pretty heavily and relied on our tests for the majority of the project.
Is there a specific piece of code you'd like feedback on? | Any and all is welcome.
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | We both agreed that we worked well together and felt that we were able to discuss ideas and plans on an equal footing. However two individual points held us back, we should have asked for help early on and agreed in the future that if we are stuck, instead of toughing it out, we should ask for help after 20 minutes have gone by.

### CODE REVIEW

File: lib/scoring.rb [hunk 1]
```diff
@@ -1,9 +1,100 @@
+require 'pry'
 module Scrabble
   class Scoring
+    LETTER_VALUES =
+    {
+      :A => 1,
+      :E => 1,
```
Comment (droberts-sea) [inline lib/scoring.rb:7]: I like this choice of data structure!

File: lib/scoring.rb [hunk 2]
```diff
@@ -1,9 +1,100 @@
...
+          word_total += 50
+        elsif
+          # checks if word length is greater then 7, and returns nil
+          user_word.length > 7 || user_word.length == 0
+          return nil
+        end
+        # iterate over each letter
+        user_word.each do |letter|
+          letter = letter.to_sym
+          if LETTER_VALUES.has_key?(letter)
```
Comment (droberts-sea) [inline lib/scoring.rb:52]: This check on line 52 is redundant! Above (line 36) you do a regex match, and here you check whether the `LETTER_VALUES` hash contains the letter. One or the other would suffice. I think I prefer the check here, since that makes it easy if you ever need to change the letters (maybe our version of scrabble starts allowing words with apostrophes).

File: lib/scoring.rb [hunk 3]
```diff
@@ -1,9 +1,100 @@
...
 
+
     def self.highest_score_from(array_of_words)
+      # RETURNS NIL FOR EMPTY ARRAY
+      if array_of_words.length == 0
+        return nil
+        # RETURNS WORD FROM ARRAY LENGTH 1
+      elsif array_of_words.length == 1
+        return array_of_words[0]
+      else
```
Comment (droberts-sea) [inline lib/scoring.rb:70]: Since both the `if` and `elsif` branches of this conditional have a return, you can omit the `else` entirely. It's not a huge difference, but it saves you some indentation.

File: specs/scoring_spec.rb [hunk 4]
```diff
@@ -42,21 +42,35 @@
...
 
     it 'returns the highest word if there are two words' do
+      array_of_words = ['cat', 'zoo']
+      Scrabble::Scoring.highest_score_from(array_of_words).must_equal 'zoo'
     end
 
     it 'if tied, prefer a word with 7 letters' do
+      array_of_words = ['dog', 'course', 'lungers']
+      Scrabble::Scoring.highest_score_from(array_of_words).must_equal 'lungers'
     end
```
Comment (droberts-sea) [inline specs/scoring_spec.rb:61]: Since `lungers` gets the 7-letter bonus, I don't think you actually have a tie here. There isn't a pair of English words that allow you to test this logic, but `zzzzzz` and `oratory` will do.

"Why bother" you ask? Well, if Miriam-Webster adds `zzzzzz` as an onomatopoeia for snoring, we need to be prepared, or you could envision a scrabble app that works across languages or that allows users to add their own words. Moreover, maybe this is the mathematician in me speaking, but I believe there's beauty and joy to be found in writing code that isn't just "good enough", but which truly solves the problem.

File: specs/scoring_spec.rb [hunk 5]
```diff
@@ -42,21 +42,35 @@
...
 
     it 'returns the only word in a length-1 array' do
+      array_of_words = ['far']
+      Scrabble::Scoring.highest_score_from(array_of_words).must_equal 'far'
     end
 
     it 'returns the highest word if there are two words' do
+      array_of_words = ['cat', 'zoo']
+      Scrabble::Scoring.highest_score_from(array_of_words).must_equal 'zoo'
```
Comment (droberts-sea) [inline specs/scoring_spec.rb:55]: It might be wise to reverse the order of this array and check that you get the same result. You would be surprised at how often this sort of thing turns up a bug. This applies to all the tiebreaker tests in this section.

File: specs/player_spec.rb [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+      # Assert
+      player.plays.must_include 'batshit'
+    end
+
+    it 'returns false if player has already won' do
+      # Arrange
+      word = "crazy"
+      word_1 = 'janky'
+      word_2 = 'coding'
+      word_3 = 'dancing'
```
Comment (droberts-sea) [inline specs/player_spec.rb:56]: Good words.

File: specs/player_spec.rb [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+      word_4 = 'hah'
+      player = Scrabble::Player.new('')
+
+      # Act
+      player.play(word)
+      player.play(word_1)
+      player.play(word_2)
+      player.play(word_3)
```
Comment (droberts-sea) [inline specs/player_spec.rb:64]: This code would be a little more clean and DRY if you stored the words in an array instead of a bunch of instance variables.

Also, I would argue that in this test, playing the first 4 words is part of the _Arrange_ step, since it's only the last word that you're interested in.

File: lib/scoring.rb [hunk 8]
```diff
@@ -1,9 +1,100 @@
...
+        array_of_words.each do |word|
+          if Scrabble::Scoring.score(word) == highest_score
+            highest_words << word
+          end
+          # binding.pry
+          # sorts through words depending on length
+          longest_word = highest_words.sort_by do |word| word.length
+            if word.length == 7
+              longest_word = word
+              return longest_word
```
Comment (droberts-sea) [inline lib/scoring.rb:87]: This loop does not select the best word. I think the core of the problem comes from the `return` statements - a `return` here doesn't return from the block, but bails out of the whole method! That means that as soon as you encounter a tie, you'll return (from the method) the first word. To demonstrate, here is a copy-paste from a pry session testing your code:

```
[9] pry(main)> Scrabble::Scoring.score('ada')
=> 4
[10] pry(main)> Scrabble::Scoring.score('ap')
=> 4
[11] pry(main)> # OK they're tied
[12] pry(main)> Scrabble::Scoring.highest_score_from(['ada', 'ap'])
=> "ada"
[13] pry(main)> Scrabble::Scoring.highest_score_from(['ap', 'ada'])
=> "ap"
```

Your implementation returns whichever word came first, when it should always prefer `ap`.

File: lib/player.rb [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  class Player
+    attr_reader :name, :played_words
+    def initialize(player_name)
+      @name = player_name
+      @played_words = []
+    end
+
+    def plays
+      return  @played_words
+    end
```
Comment (droberts-sea) [inline lib/player.rb:13]: Another way to approach this method would be to name the instance variable `@plays` and use an `attr_reader`.

File: lib/player.rb [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    # calculates the score of all the words
+    def total_score
+      word_scores = @played_words.map do |word|
+         Scrabble::Scoring.score(word)
+      end
+      # this will allow me to see if something inside a loop is not what I expect
+      # if word_scores.include?(nil)
+      #   binding.pry
+      # end
+      return word_scores.inject(:+)
```
Comment (droberts-sea) [inline lib/player.rb:34]: Beautiful!

File: lib/player.rb [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    def won
+      if total_score > 100
+        return true
+      else
+        return false
+
+      end
+
+      def highest_scoring_word
+        Scrabble::Scoring.highest_score_from(plays)
```
Comment (droberts-sea) [inline lib/player.rb:46]: You're missing an end here - `highest_scoring_word` is defined _inside_ of `won`. Turns out this is valid Ruby syntax, but it'd probably not what you wanted.

File: lib/player.rb [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+      #   binding.pry
+      # end
+      return word_scores.inject(:+)
+    end
+
+    def won
+      if total_score > 100
+        return true
+      else
+        return false
```
Comment (droberts-sea) [inline lib/player.rb:41]: Since `>` will always return `true` or `false`, this could be shortened to:

```ruby
def won?
  return total_score > 100
end
```

File: specs/player_spec.rb [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+
+    xit '' do
+
+    end
+  end
+
+  describe 'won?' do
+    it 'if player has > 100 point score, return true' do
+      word = "crazy"
+      word_1 = 'janky'
```
Comment (droberts-sea) [inline specs/player_spec.rb:113]: Here are some other questions I would like to see answered by these tests:

- What if the player has not played any words at all?
- What if the player has < 100 points?
- What if the player has exactly 100 points?

File: specs/player_spec.rb [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+      player.play(word)
+      player.play(word_1)
+      player.play(word_2)
+      player.play(word_3)
+      player.plays
+      player.total_score
+
+      # Assert
+      player.total_score.must_be Integer
+
```
Comment (droberts-sea) [inline specs/player_spec.rb:101]: You should also calculate the score and check the result.

**Comment by droberts-sea** [comment]
# Scrabble
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | yes
Both Teammates contributed to the codebase | yes
Regular Commits with meaningful commit messages | yes
**Scoring class** |
`score` method | yes
Uses appropriate data structure to store the letter score | yes
Appropriately handles edge cases | yes
`highest_score_from` method | yes
Appropriately handles edge cases | tie breaking doesn't work - see inline
Test implementations for `highest_score_from`| see inline
**Player class** |
Uses an array to keep track of words played | yes
Uses existing code to create `highest_scoring_word` and `highest_word_score` | some - methods have othr problems
Returns `true` or `false` from `won?` method | yes
Tests edge cases on the `play(word)` | some - see inline
**TileBag class** |
Uses a data structure to keep the set of default tiles in the bag | N/A
`draw_tiles` method uses the parameter to draw a set of random tiles | N/A
`tiles_remaining` method returns a count | N/A
Tests `draw_tiles` edge cases | N/A

The big learning goals I'm looking to see demonstrated on this project are:

- Building complex logic (tie-breaking logic in `highest_score_from`)
- Test coverage
- Object composition (interaction between `Player` and `TileBag`)

You've made a good start on all of these, but there are substantial pieces missing for each. Please review my inline comments, and try and focus on these learning goals for Ride Share and next week's Hotel project. I'll be excited to see what you come up with. I appreciate that both of you have been putting in the hours recently - keep up the hard work and dedication!
